<font size=5><b> Qs6.	Bonus (5 Points): Use the 3 most successful techniques espoused in question 3 to another NetFlow dataset.

# Cleaning the new dataset

In [ ]:
# Please note!!!  The group had to expand the RAM available to Jupyter for this code to run. 
# We exapnded the RAM to 4GB and recommend you do the same.  Please see the answer to the below
# stackoverflow page concerning expanding the amount of max_buffer_size. 
# https://stackoverflow.com/questions/57948003/how-to-increase-jupyter-notebook-memory-limit

In [2]:
#Importing packages
import pandas as pd
import warnings

warnings.filterwarnings('ignore')

In [3]:
# Please note that the extra credit dataset is too large to be sent without modification.  You will need to use 7-zip or
# an equivalent application to convert the .xz file. 

In [4]:
#Reading df
df = pd.read_csv(r'capture20110810.binetflow')
df.head()

StartTime       Dur Proto        SrcAddr Sport    Dir  \
0  2011/08/10 09:46:59.607825  1.026539   tcp  94.44.127.113  1577     ->   
1  2011/08/10 09:47:00.634364  1.009595   tcp  94.44.127.113  1577     ->   
2  2011/08/10 09:47:48.185538  3.056586   tcp   147.32.86.89  4768     ->   
3  2011/08/10 09:47:48.230897  3.111769   tcp   147.32.86.89  4788     ->   
4  2011/08/10 09:47:48.963351  3.083411   tcp   147.32.86.89  4850     ->   

        DstAddr Dport State  sTos  dTos  TotPkts  TotBytes  SrcBytes  \
0  147.32.84.59  6881  S_RA   0.0   0.0        4       276       156   
1  147.32.84.59  6881  S_RA   0.0   0.0        4       276       156   
2   77.75.73.33    80  SR_A   0.0   0.0        3       182       122   
3   77.75.73.33    80  SR_A   0.0   0.0        3       182       122   
4   77.75.73.33    80  SR_A   0.0   0.0        3       182       122   

                                    Label  
0  flow=Background-Established-cmpgw-CVUT  
1  flow=Background-Established-cmpgw-CVUT  
2             flow=Background-TCP-Attempt  
3             flow=Background-TCP-Attempt  
4             flow=Background-TCP-Attempt

In [3]:
#Dropping unneccesary columns
df.drop(['StartTime'],axis=1,inplace = True)

In [4]:
df.State.nunique()

230

In [5]:
df.head()

Dur Proto        SrcAddr Sport    Dir       DstAddr Dport State  sTos  \
0  1.026539   tcp  94.44.127.113  1577     ->  147.32.84.59  6881  S_RA   0.0   
1  1.009595   tcp  94.44.127.113  1577     ->  147.32.84.59  6881  S_RA   0.0   
2  3.056586   tcp   147.32.86.89  4768     ->   77.75.73.33    80  SR_A   0.0   
3  3.111769   tcp   147.32.86.89  4788     ->   77.75.73.33    80  SR_A   0.0   
4  3.083411   tcp   147.32.86.89  4850     ->   77.75.73.33    80  SR_A   0.0   

   dTos  TotPkts  TotBytes  SrcBytes                                   Label  
0   0.0        4       276       156  flow=Background-Established-cmpgw-CVUT  
1   0.0        4       276       156  flow=Background-Established-cmpgw-CVUT  
2   0.0        3       182       122             flow=Background-TCP-Attempt  
3   0.0        3       182       122             flow=Background-TCP-Attempt  
4   0.0        3       182       122             flow=Background-TCP-Attempt

In [6]:
#Renaming the columns using their proper names from the previous dataset
df.rename(columns = {list(df)[0]:'Duration',list(df)[1]:'Protocol',list(df)[2]:'Source_IP',list(df)[3]:'Source_Port',
                    list(df)[5]:'Destination_IP',list(df)[6]:'Destination_Port',list(df)[7]:'Flags',list(df)[8]:'Source_Type_of_Service',
                    list(df)[9]:'Dest_Type_of_Service',list(df)[10]:'Packets',list(df)[11]:'Total_Bytes',list(df)[12]:'Source_Bytes'}, inplace = True)

In [7]:
df.head()

Duration Protocol      Source_IP Source_Port    Dir Destination_IP  \
0  1.026539      tcp  94.44.127.113        1577     ->   147.32.84.59   
1  1.009595      tcp  94.44.127.113        1577     ->   147.32.84.59   
2  3.056586      tcp   147.32.86.89        4768     ->    77.75.73.33   
3  3.111769      tcp   147.32.86.89        4788     ->    77.75.73.33   
4  3.083411      tcp   147.32.86.89        4850     ->    77.75.73.33   

  Destination_Port Flags  Source_Type_of_Service  Dest_Type_of_Service  \
0             6881  S_RA                     0.0                   0.0   
1             6881  S_RA                     0.0                   0.0   
2               80  SR_A                     0.0                   0.0   
3               80  SR_A                     0.0                   0.0   
4               80  SR_A                     0.0                   0.0   

   Packets  Total_Bytes  Source_Bytes                                   Label  
0        4          276           156  flow=Background-Established-cmpgw-CVUT  
1        4          276           156  flow=Background-Established-cmpgw-CVUT  
2        3          182           122             flow=Background-TCP-Attempt  
3        3          182           122             flow=Background-TCP-Attempt  
4        3          182           122             flow=Background-TCP-Attempt

In [8]:
df.Dir.unique()

array(['   ->', '   ?>', '  <->', '  <?>', '  who', '  <-', '  <?'],
      dtype=object)

In [9]:
direction = ['   ->','  <->','  <-']
df = df.loc[df['Dir'].isin(direction)]

In [10]:
df #Now 2.813Million rows - Only about 10,000 values removed from 2.824Million rows

Duration Protocol        Source_IP Source_Port    Dir  \
0          1.026539      tcp    94.44.127.113        1577     ->   
1          1.009595      tcp    94.44.127.113        1577     ->   
2          3.056586      tcp     147.32.86.89        4768     ->   
3          3.111769      tcp     147.32.86.89        4788     ->   
4          3.083411      tcp     147.32.86.89        4850     ->   
5          3.097288      tcp     147.32.86.89        4866     ->   
6          1.048908      tcp  213.200.244.217       47908     ->   
7          4.373526      tcp     75.105.28.60        1419     ->   
8          4.827912      tcp     75.105.28.60        1491     ->   
9          0.049697      tcp   178.111.79.115       41752     ->   
10       328.361664      tcp     147.32.84.59       49185     ->   
11         5.242459      tcp     75.105.28.60        1586     ->   
12         0.972390      tcp     89.31.40.106       28451     ->   
13         0.923098      tcp     89.31.40.106       13717     ->   
14         1.009763      tcp    188.112.70.72        1817     ->   
15         0.969967      tcp     85.248.56.40       42480     ->   
16         2.853907      tcp   41.188.145.202        2285     ->   
17         2.948305      tcp   41.188.145.202        2288     ->   
18         2.956415      tcp   41.188.145.202        2291     ->   
19         0.920374      tcp     89.31.40.106       39927     ->   
20         6.031078      tcp   41.188.145.202        2285     ->   
21         6.035653      tcp   41.188.145.202        2288     ->   
22         6.028175      tcp   41.188.145.202        2291     ->   
24         1.383731      tcp   213.24.237.172       15007     ->   
25         0.939583      tcp     85.248.56.40       42572     ->   
26         3.015055      tcp    69.231.198.54        4144     ->   
27         0.987838      tcp    188.112.70.72        2057     ->   
28         1.419550      tcp    95.153.177.13       35049     ->   
29         3.024265      tcp    69.231.198.54        4154     ->   
30         0.990760      tcp     85.248.56.40       42650     ->   
...             ...      ...              ...         ...    ...   
2824606    0.000767      udp    83.46.238.157       48002    <->   
2824607    0.000373      udp     147.32.84.59       60428    <->   
2824608    0.000947      udp    98.87.173.219       28053    <->   
2824609    0.000333      udp     147.32.84.59       57875    <->   
2824610    0.002661      udp       80.214.5.1       26863    <->   
2824611    0.000870      udp     149.5.45.254       43061    <->   
2824612    0.000800      udp      94.113.4.90        4858    <->   
2824613    0.021536      tcp    66.249.66.241       40474     ->   
2824614    0.000233      udp    147.32.84.138       40049    <->   
2824615    0.000256      udp    147.32.84.138       47007    <->   
2824616    0.000299      udp    147.32.84.138       53567    <->   
2824617    0.000303      udp    147.32.84.138       42937    <->   
2824618    0.001034      udp     84.54.188.23        7198    <->   
2824619    0.000318      udp    147.32.84.138       56632    <->   
2824620    0.000138      udp    147.32.84.138       54724    <->   
2824621    0.000209      udp    147.32.84.138       53255    <->   
2824622    0.000253      udp    147.32.84.138       40427    <->   
2824623    0.001018      udp  130.243.190.199       25847    <->   
2824624    0.000144      udp    147.32.84.138       58080    <->   
2824625    0.000271      udp    147.32.84.138       52130    <->   
2824626    0.000207      udp    147.32.84.138       35552    <->   
2824627    0.000207      udp    147.32.84.138       58286    <->   
2824628    0.000858      udp    178.40.143.31       18403    <->   
2824629    0.000861      udp     88.222.4.220       53717    <->   
2824630    0.000330      udp     147.32.86.92       36363    <->   
2824631    0.000393      udp     147.32.86.92       36363    <->   
2824632    0.000935      udp     58.165.41.84       60122    <->   

In [11]:
df.replace(['   ->','  <->','  <-'],  
                            ['outgoing','two-way','incoming'],inplace=True)
# df.replace({'Dir':{'   ->': 'outgoing', '  <->': 'two-way', '  <-': 'incoming'}})

In [12]:
df.Dir.unique()

array(['outgoing', 'two-way', 'incoming'], dtype=object)

In [13]:
df.rename(columns = {list(df)[4]:'Direction'}, inplace = True)

In [14]:
df

Duration Protocol        Source_IP Source_Port Direction  \
0          1.026539      tcp    94.44.127.113        1577  outgoing   
1          1.009595      tcp    94.44.127.113        1577  outgoing   
2          3.056586      tcp     147.32.86.89        4768  outgoing   
3          3.111769      tcp     147.32.86.89        4788  outgoing   
4          3.083411      tcp     147.32.86.89        4850  outgoing   
5          3.097288      tcp     147.32.86.89        4866  outgoing   
6          1.048908      tcp  213.200.244.217       47908  outgoing   
7          4.373526      tcp     75.105.28.60        1419  outgoing   
8          4.827912      tcp     75.105.28.60        1491  outgoing   
9          0.049697      tcp   178.111.79.115       41752  outgoing   
10       328.361664      tcp     147.32.84.59       49185  outgoing   
11         5.242459      tcp     75.105.28.60        1586  outgoing   
12         0.972390      tcp     89.31.40.106       28451  outgoing   
13         0.923098      tcp     89.31.40.106       13717  outgoing   
14         1.009763      tcp    188.112.70.72        1817  outgoing   
15         0.969967      tcp     85.248.56.40       42480  outgoing   
16         2.853907      tcp   41.188.145.202        2285  outgoing   
17         2.948305      tcp   41.188.145.202        2288  outgoing   
18         2.956415      tcp   41.188.145.202        2291  outgoing   
19         0.920374      tcp     89.31.40.106       39927  outgoing   
20         6.031078      tcp   41.188.145.202        2285  outgoing   
21         6.035653      tcp   41.188.145.202        2288  outgoing   
22         6.028175      tcp   41.188.145.202        2291  outgoing   
24         1.383731      tcp   213.24.237.172       15007  outgoing   
25         0.939583      tcp     85.248.56.40       42572  outgoing   
26         3.015055      tcp    69.231.198.54        4144  outgoing   
27         0.987838      tcp    188.112.70.72        2057  outgoing   
28         1.419550      tcp    95.153.177.13       35049  outgoing   
29         3.024265      tcp    69.231.198.54        4154  outgoing   
30         0.990760      tcp     85.248.56.40       42650  outgoing   
...             ...      ...              ...         ...       ...   
2824606    0.000767      udp    83.46.238.157       48002   two-way   
2824607    0.000373      udp     147.32.84.59       60428   two-way   
2824608    0.000947      udp    98.87.173.219       28053   two-way   
2824609    0.000333      udp     147.32.84.59       57875   two-way   
2824610    0.002661      udp       80.214.5.1       26863   two-way   
2824611    0.000870      udp     149.5.45.254       43061   two-way   
2824612    0.000800      udp      94.113.4.90        4858   two-way   
2824613    0.021536      tcp    66.249.66.241       40474  outgoing   
2824614    0.000233      udp    147.32.84.138       40049   two-way   
2824615    0.000256      udp    147.32.84.138       47007   two-way   
2824616    0.000299      udp    147.32.84.138       53567   two-way   
2824617    0.000303      udp    147.32.84.138       42937   two-way   
2824618    0.001034      udp     84.54.188.23        7198   two-way   
2824619    0.000318      udp    147.32.84.138       56632   two-way   
2824620    0.000138      udp    147.32.84.138       54724   two-way   
2824621    0.000209      udp    147.32.84.138       53255   two-way   
2824622    0.000253      udp    147.32.84.138       40427   two-way   
2824623    0.001018      udp  130.243.190.199       25847   two-way   
2824624    0.000144      udp    147.32.84.138       58080   two-way   
2824625    0.000271      udp    147.32.84.138       52130   two-way   
2824626    0.000207      udp    147.32.84.138       35552   two-way   
2824627    0.000207      udp    147.32.84.138       58286   two-way   
2824628    0.000858      udp    178.40.143.31       18403   two-way   
2824629    0.000861      udp     88.222.4.220       53717   two-way   
2824630    0.000330      udp     14

In [15]:
df.Label.unique()

array(['flow=Background-Established-cmpgw-CVUT',
       'flow=Background-TCP-Attempt', 'flow=Background-TCP-Established',
       'flow=To-Background-CVUT-Proxy', 'flow=Background-UDP-Established',
       'flow=To-Background-UDP-CVUT-DNS-Server',
       'flow=Background-UDP-Attempt',
       'flow=From-Normal-V42-UDP-CVUT-DNS-Server',
       'flow=Background-UDP-NTP-Established-1',
       'flow=To-Background-CVUT-WebServer',
       'flow=Background-Attempt-cmpgw-CVUT', 'flow=Background',
       'flow=From-Background-CVUT-Proxy',
       'flow=Background-google-analytics3', 'flow=From-Normal-V42-Jist',
       'flow=Background-google-analytics14',
       'flow=Background-ajax.google', 'flow=Background-google-analytics6',
       'flow=Background-google-analytics13', 'flow=Background-google-pop',
       'flow=To-Normal-V42-UDP-NTP-server',
       'flow=From-Normal-V42-Stribrek',
       'flow=Background-google-analytics1',
       'flow=Background-google-analytics16',
       'flow=Background-go

In [16]:
df['Label'] = df['Label'].str.replace(r'flow=', '')

In [17]:
df['Label'] = df['Label'].str.replace(r'From-', '')

In [18]:
df['Label'] = df['Label'].str.replace(r'To-', '')

In [19]:
df['Label'] = df['Label'].str.extract(r'(\w+)')

In [20]:
df.Label.unique()

array(['Background', 'Normal', 'Botnet'], dtype=object)

In [21]:
df

Duration Protocol        Source_IP Source_Port Direction  \
0          1.026539      tcp    94.44.127.113        1577  outgoing   
1          1.009595      tcp    94.44.127.113        1577  outgoing   
2          3.056586      tcp     147.32.86.89        4768  outgoing   
3          3.111769      tcp     147.32.86.89        4788  outgoing   
4          3.083411      tcp     147.32.86.89        4850  outgoing   
5          3.097288      tcp     147.32.86.89        4866  outgoing   
6          1.048908      tcp  213.200.244.217       47908  outgoing   
7          4.373526      tcp     75.105.28.60        1419  outgoing   
8          4.827912      tcp     75.105.28.60        1491  outgoing   
9          0.049697      tcp   178.111.79.115       41752  outgoing   
10       328.361664      tcp     147.32.84.59       49185  outgoing   
11         5.242459      tcp     75.105.28.60        1586  outgoing   
12         0.972390      tcp     89.31.40.106       28451  outgoing   
13         0.923098      tcp     89.31.40.106       13717  outgoing   
14         1.009763      tcp    188.112.70.72        1817  outgoing   
15         0.969967      tcp     85.248.56.40       42480  outgoing   
16         2.853907      tcp   41.188.145.202        2285  outgoing   
17         2.948305      tcp   41.188.145.202        2288  outgoing   
18         2.956415      tcp   41.188.145.202        2291  outgoing   
19         0.920374      tcp     89.31.40.106       39927  outgoing   
20         6.031078      tcp   41.188.145.202        2285  outgoing   
21         6.035653      tcp   41.188.145.202        2288  outgoing   
22         6.028175      tcp   41.188.145.202        2291  outgoing   
24         1.383731      tcp   213.24.237.172       15007  outgoing   
25         0.939583      tcp     85.248.56.40       42572  outgoing   
26         3.015055      tcp    69.231.198.54        4144  outgoing   
27         0.987838      tcp    188.112.70.72        2057  outgoing   
28         1.419550      tcp    95.153.177.13       35049  outgoing   
29         3.024265      tcp    69.231.198.54        4154  outgoing   
30         0.990760      tcp     85.248.56.40       42650  outgoing   
...             ...      ...              ...         ...       ...   
2824606    0.000767      udp    83.46.238.157       48002   two-way   
2824607    0.000373      udp     147.32.84.59       60428   two-way   
2824608    0.000947      udp    98.87.173.219       28053   two-way   
2824609    0.000333      udp     147.32.84.59       57875   two-way   
2824610    0.002661      udp       80.214.5.1       26863   two-way   
2824611    0.000870      udp     149.5.45.254       43061   two-way   
2824612    0.000800      udp      94.113.4.90        4858   two-way   
2824613    0.021536      tcp    66.249.66.241       40474  outgoing   
2824614    0.000233      udp    147.32.84.138       40049   two-way   
2824615    0.000256      udp    147.32.84.138       47007   two-way   
2824616    0.000299      udp    147.32.84.138       53567   two-way   
2824617    0.000303      udp    147.32.84.138       42937   two-way   
2824618    0.001034      udp     84.54.188.23        7198   two-way   
2824619    0.000318      udp    147.32.84.138       56632   two-way   
2824620    0.000138      udp    147.32.84.138       54724   two-way   
2824621    0.000209      udp    147.32.84.138       53255   two-way   
2824622    0.000253      udp    147.32.84.138       40427   two-way   
2824623    0.001018      udp  130.243.190.199       25847   two-way   
2824624    0.000144      udp    147.32.84.138       58080   two-way   
2824625    0.000271      udp    147.32.84.138       52130   two-way   
2824626    0.000207      udp    147.32.84.138       35552   two-way   
2824627    0.000207      udp    147.32.84.138       58286   two-way   
2824628    0.000858      udp    178.40.143.31       18403   two-way   
2824629    0.000861      udp     88.222.4.220       53717   two-way   
2824630    0.000330      udp     14

In [22]:
df.dropna(inplace=True)


In [23]:
df #dropping all null values removes 200k records = 2.613Million rows

Duration Protocol        Source_IP Source_Port Direction  \
0          1.026539      tcp    94.44.127.113        1577  outgoing   
1          1.009595      tcp    94.44.127.113        1577  outgoing   
2          3.056586      tcp     147.32.86.89        4768  outgoing   
3          3.111769      tcp     147.32.86.89        4788  outgoing   
4          3.083411      tcp     147.32.86.89        4850  outgoing   
5          3.097288      tcp     147.32.86.89        4866  outgoing   
6          1.048908      tcp  213.200.244.217       47908  outgoing   
7          4.373526      tcp     75.105.28.60        1419  outgoing   
8          4.827912      tcp     75.105.28.60        1491  outgoing   
9          0.049697      tcp   178.111.79.115       41752  outgoing   
10       328.361664      tcp     147.32.84.59       49185  outgoing   
11         5.242459      tcp     75.105.28.60        1586  outgoing   
12         0.972390      tcp     89.31.40.106       28451  outgoing   
13         0.923098      tcp     89.31.40.106       13717  outgoing   
14         1.009763      tcp    188.112.70.72        1817  outgoing   
15         0.969967      tcp     85.248.56.40       42480  outgoing   
16         2.853907      tcp   41.188.145.202        2285  outgoing   
17         2.948305      tcp   41.188.145.202        2288  outgoing   
18         2.956415      tcp   41.188.145.202        2291  outgoing   
19         0.920374      tcp     89.31.40.106       39927  outgoing   
20         6.031078      tcp   41.188.145.202        2285  outgoing   
21         6.035653      tcp   41.188.145.202        2288  outgoing   
22         6.028175      tcp   41.188.145.202        2291  outgoing   
24         1.383731      tcp   213.24.237.172       15007  outgoing   
25         0.939583      tcp     85.248.56.40       42572  outgoing   
26         3.015055      tcp    69.231.198.54        4144  outgoing   
27         0.987838      tcp    188.112.70.72        2057  outgoing   
28         1.419550      tcp    95.153.177.13       35049  outgoing   
29         3.024265      tcp    69.231.198.54        4154  outgoing   
30         0.990760      tcp     85.248.56.40       42650  outgoing   
...             ...      ...              ...         ...       ...   
2824605    0.153318      tcp     147.32.86.20        4161  outgoing   
2824606    0.000767      udp    83.46.238.157       48002   two-way   
2824607    0.000373      udp     147.32.84.59       60428   two-way   
2824608    0.000947      udp    98.87.173.219       28053   two-way   
2824609    0.000333      udp     147.32.84.59       57875   two-way   
2824610    0.002661      udp       80.214.5.1       26863   two-way   
2824611    0.000870      udp     149.5.45.254       43061   two-way   
2824612    0.000800      udp      94.113.4.90        4858   two-way   
2824613    0.021536      tcp    66.249.66.241       40474  outgoing   
2824614    0.000233      udp    147.32.84.138       40049   two-way   
2824615    0.000256      udp    147.32.84.138       47007   two-way   
2824616    0.000299      udp    147.32.84.138       53567   two-way   
2824617    0.000303      udp    147.32.84.138       42937   two-way   
2824618    0.001034      udp     84.54.188.23        7198   two-way   
2824619    0.000318      udp    147.32.84.138       56632   two-way   
2824620    0.000138      udp    147.32.84.138       54724   two-way   
2824621    0.000209      udp    147.32.84.138       53255   two-way   
2824622    0.000253      udp    147.32.84.138       40427   two-way   
2824623    0.001018      udp  130.243.190.199       25847   two-way   
2824624    0.000144      udp    147.32.84.138       58080   two-way   
2824625    0.000271      udp    147.32.84.138       52130   two-way   
2824626    0.000207      udp    147.32.84.138       35552   two-way   
2824627    0.000207      udp    147.32.84.138       58286   two-way   
2824628    0.000858      udp    178.40.143.31       18403   two-way   
2824629    0.000861      udp     88

In [24]:
df['Source_Type_of_Service'].unique()

array([0., 3., 2., 1.])

In [25]:
df['Dest_Type_of_Service'].unique()

array([0., 2., 3., 1.])

# Label Encoding all Categorical Variables 

### Direction

In [26]:
# Import label encoder 
from sklearn import preprocessing 
  
# label_encoder object knows how to understand word labels. 
label_encoder = preprocessing.LabelEncoder() 
  
# Encode labels in column 'species'. 
cat_dir_data= label_encoder.fit_transform(df[['Direction']]) 
New_cat_dir = pd.DataFrame(cat_dir_data)
New_cat_dir=New_cat_dir.rename(columns={0:'Direction'})


In [27]:
New_cat_dir.Direction.nunique()

3

### Protocol

In [28]:
df.Protocol.unique()

array(['tcp', 'udp', 'rtp', 'icmp', 'rtcp', 'udt'], dtype=object)

In [29]:
label_encoder = preprocessing.LabelEncoder() 
  
# Encode labels in column 'species'. 
Protocol_Copy= label_encoder.fit_transform(df[['Protocol']]) 
New_cat_protocol = pd.DataFrame(Protocol_Copy)
New_cat_protocol=New_cat_protocol.rename(columns={0:'Protocol'})

In [30]:
New_cat_protocol.Protocol.nunique()

6

### Flags

In [31]:
df.Flags.nunique()

169

In [32]:
label_encoder = preprocessing.LabelEncoder() 
  
# Encode labels in column 'species'. 
cat_flags_data= label_encoder.fit_transform(df[['Flags']]) 
New_cat_flags = pd.DataFrame(cat_flags_data)
New_cat_flags=New_cat_flags.rename(columns={0:'Flags'})

In [33]:
New_cat_flags.nunique()

Flags    169
dtype: int64

### Source Country

In [34]:
# Now recreating the geoip2 data. 

#Source country
import pandas as pd
from geolite2 import geolite2

def get_country(ip):
    try:
        df = geo.get(ip)
    except ValueError:
        return pd.np.nan
    try:
        return df['country']['names']['en'] if df else pd.np.nan
    except KeyError:
        return pd.np.nan

geo = geolite2.reader()
# get unique IPs
unique_ips = df['Source_IP'].unique()
# make series out of it
unique_ips = pd.Series(unique_ips, index = unique_ips)
# map IP --> country
df['country'] = df['Source_IP'].map(unique_ips.apply(get_country))

geolite2.close()

In [35]:
df.country.value_counts() #Again Czech Republic has the highest count

Czechia                               1478951
United States                          165122
Bulgaria                               122854
Russia                                  79603
United Kingdom                          77600
Germany                                 47334
Australia                               41318
Japan                                   35162
Italy                                   30903
Greece                                  29976
Taiwan                                  28216
Ukraine                                 25289
China                                   25019
Netherlands                             22619
France                                  22171
Canada                                  21398
Spain                                   18317
Brazil                                  16866
Romania                                 15106
Austria                                 13311
India                                   11674
Republic of Lithuania             

In [36]:
df.dropna(inplace=True)

In [37]:
df #Another 10k removed

Duration Protocol        Source_IP Source_Port Direction  \
0          1.026539      tcp    94.44.127.113        1577  outgoing   
1          1.009595      tcp    94.44.127.113        1577  outgoing   
2          3.056586      tcp     147.32.86.89        4768  outgoing   
3          3.111769      tcp     147.32.86.89        4788  outgoing   
4          3.083411      tcp     147.32.86.89        4850  outgoing   
5          3.097288      tcp     147.32.86.89        4866  outgoing   
6          1.048908      tcp  213.200.244.217       47908  outgoing   
7          4.373526      tcp     75.105.28.60        1419  outgoing   
8          4.827912      tcp     75.105.28.60        1491  outgoing   
9          0.049697      tcp   178.111.79.115       41752  outgoing   
10       328.361664      tcp     147.32.84.59       49185  outgoing   
11         5.242459      tcp     75.105.28.60        1586  outgoing   
12         0.972390      tcp     89.31.40.106       28451  outgoing   
13         0.923098      tcp     89.31.40.106       13717  outgoing   
14         1.009763      tcp    188.112.70.72        1817  outgoing   
15         0.969967      tcp     85.248.56.40       42480  outgoing   
16         2.853907      tcp   41.188.145.202        2285  outgoing   
17         2.948305      tcp   41.188.145.202        2288  outgoing   
18         2.956415      tcp   41.188.145.202        2291  outgoing   
19         0.920374      tcp     89.31.40.106       39927  outgoing   
20         6.031078      tcp   41.188.145.202        2285  outgoing   
21         6.035653      tcp   41.188.145.202        2288  outgoing   
22         6.028175      tcp   41.188.145.202        2291  outgoing   
24         1.383731      tcp   213.24.237.172       15007  outgoing   
25         0.939583      tcp     85.248.56.40       42572  outgoing   
26         3.015055      tcp    69.231.198.54        4144  outgoing   
27         0.987838      tcp    188.112.70.72        2057  outgoing   
28         1.419550      tcp    95.153.177.13       35049  outgoing   
29         3.024265      tcp    69.231.198.54        4154  outgoing   
30         0.990760      tcp     85.248.56.40       42650  outgoing   
...             ...      ...              ...         ...       ...   
2824605    0.153318      tcp     147.32.86.20        4161  outgoing   
2824606    0.000767      udp    83.46.238.157       48002   two-way   
2824607    0.000373      udp     147.32.84.59       60428   two-way   
2824608    0.000947      udp    98.87.173.219       28053   two-way   
2824609    0.000333      udp     147.32.84.59       57875   two-way   
2824610    0.002661      udp       80.214.5.1       26863   two-way   
2824611    0.000870      udp     149.5.45.254       43061   two-way   
2824612    0.000800      udp      94.113.4.90        4858   two-way   
2824613    0.021536      tcp    66.249.66.241       40474  outgoing   
2824614    0.000233      udp    147.32.84.138       40049   two-way   
2824615    0.000256      udp    147.32.84.138       47007   two-way   
2824616    0.000299      udp    147.32.84.138       53567   two-way   
2824617    0.000303      udp    147.32.84.138       42937   two-way   
2824618    0.001034      udp     84.54.188.23        7198   two-way   
2824619    0.000318      udp    147.32.84.138       56632   two-way   
2824620    0.000138      udp    147.32.84.138       54724   two-way   
2824621    0.000209      udp    147.32.84.138       53255   two-way   
2824622    0.000253      udp    147.32.84.138       40427   two-way   
2824623    0.001018      udp  130.243.190.199       25847   two-way   
2824624    0.000144      udp    147.32.84.138       58080   two-way   
2824625    0.000271      udp    147.32.84.138       52130   two-way   
2824626    0.000207      udp    147.32.84.138       35552   two-way   
2824627    0.000207      udp    147.32.84.138       58286   two-way   
2824628    0.000858      udp    178.40.143.31       18403   two-way   
2824629    0.000861      udp     88

### Source City

In [38]:
#source city
def get_city(ip):
    try:
        df = geo.get(ip)
    except ValueError:
        return pd.np.nan
    try:
        return df['city']['names']['en'] if df else pd.np.nan
    except KeyError:
        return pd.np.nan

geo = geolite2.reader()
# get unique IPs
unique_ips = df['Source_IP'].unique()
# make series out of it
unique_ips = pd.Series(unique_ips, index = unique_ips)
# map IP --> country
df['city'] = df['Source_IP'].map(unique_ips.apply(get_city))

geolite2.close()

In [39]:
df.city.value_counts()

Prague                 1459840
San Antonio              44938
Sofia                    26005
Moscow                   17671
Taipei                   12717
Athens                   11967
Tokyo                     9739
Varna                     9315
Plovdiv                   9308
Mountain View             8971
Singapore                 6971
St Petersburg             6484
Tbilisi                   5915
Burgas                    5786
Kiev                      5741
Guangzhou                 5342
Chisinau                  4394
Berlin                    4318
Bucharest                 4132
London                    4034
Stara Zagora              3989
Riga                      3915
Vilnius                   3700
Sydney                    3595
Vienna                    3522
Thessaloniki              3400
Central                   3351
Beijing                   3249
Shanghai                  3185
Toronto                   3030
                        ...   
St. Margrethen               1
Nuevo Ba

In [40]:
df.dropna(inplace=True)

In [41]:
df #About 200k values removed

Duration Protocol        Source_IP Source_Port Direction  \
0          1.026539      tcp    94.44.127.113        1577  outgoing   
1          1.009595      tcp    94.44.127.113        1577  outgoing   
2          3.056586      tcp     147.32.86.89        4768  outgoing   
3          3.111769      tcp     147.32.86.89        4788  outgoing   
4          3.083411      tcp     147.32.86.89        4850  outgoing   
5          3.097288      tcp     147.32.86.89        4866  outgoing   
6          1.048908      tcp  213.200.244.217       47908  outgoing   
7          4.373526      tcp     75.105.28.60        1419  outgoing   
8          4.827912      tcp     75.105.28.60        1491  outgoing   
9          0.049697      tcp   178.111.79.115       41752  outgoing   
10       328.361664      tcp     147.32.84.59       49185  outgoing   
11         5.242459      tcp     75.105.28.60        1586  outgoing   
12         0.972390      tcp     89.31.40.106       28451  outgoing   
13         0.923098      tcp     89.31.40.106       13717  outgoing   
14         1.009763      tcp    188.112.70.72        1817  outgoing   
15         0.969967      tcp     85.248.56.40       42480  outgoing   
16         2.853907      tcp   41.188.145.202        2285  outgoing   
17         2.948305      tcp   41.188.145.202        2288  outgoing   
18         2.956415      tcp   41.188.145.202        2291  outgoing   
19         0.920374      tcp     89.31.40.106       39927  outgoing   
20         6.031078      tcp   41.188.145.202        2285  outgoing   
21         6.035653      tcp   41.188.145.202        2288  outgoing   
22         6.028175      tcp   41.188.145.202        2291  outgoing   
25         0.939583      tcp     85.248.56.40       42572  outgoing   
26         3.015055      tcp    69.231.198.54        4144  outgoing   
27         0.987838      tcp    188.112.70.72        2057  outgoing   
28         1.419550      tcp    95.153.177.13       35049  outgoing   
29         3.024265      tcp    69.231.198.54        4154  outgoing   
30         0.990760      tcp     85.248.56.40       42650  outgoing   
31         2.923240      tcp    69.231.198.54        4168  outgoing   
...             ...      ...              ...         ...       ...   
2824602    0.000761      udp     94.26.14.157       63035   two-way   
2824603    0.108535      tcp    66.249.72.141       62493  outgoing   
2824604    0.000290      udp     147.32.86.20       65160   two-way   
2824605    0.153318      tcp     147.32.86.20        4161  outgoing   
2824606    0.000767      udp    83.46.238.157       48002   two-way   
2824607    0.000373      udp     147.32.84.59       60428   two-way   
2824609    0.000333      udp     147.32.84.59       57875   two-way   
2824610    0.002661      udp       80.214.5.1       26863   two-way   
2824612    0.000800      udp      94.113.4.90        4858   two-way   
2824613    0.021536      tcp    66.249.66.241       40474  outgoing   
2824614    0.000233      udp    147.32.84.138       40049   two-way   
2824615    0.000256      udp    147.32.84.138       47007   two-way   
2824616    0.000299      udp    147.32.84.138       53567   two-way   
2824617    0.000303      udp    147.32.84.138       42937   two-way   
2824618    0.001034      udp     84.54.188.23        7198   two-way   
2824619    0.000318      udp    147.32.84.138       56632   two-way   
2824620    0.000138      udp    147.32.84.138       54724   two-way   
2824621    0.000209      udp    147.32.84.138       53255   two-way   
2824622    0.000253      udp    147.32.84.138       40427   two-way   
2824623    0.001018      udp  130.243.190.199       25847   two-way   
2824624    0.000144      udp    147.32.84.138       58080   two-way   
2824625    0.000271      udp    147.32.84.138       52130   two-way   
2824626    0.000207      udp    147.32.84.138       35552   two-way   
2824627    0.000207      udp    147.32.84.138       58286   two-way   
2824628    0.000858      udp    178

In [42]:
label_encoder = preprocessing.LabelEncoder() 
  
# Encode labels in column 'species'. 
cat_Src_City= label_encoder.fit_transform(df[['city']]) 
New_cat_Src_City = pd.DataFrame(cat_Src_City)
New_cat_Src_City= New_cat_Src_City.rename(columns={0:'Src_city'})

In [43]:
New_cat_Src_City.nunique()

Src_city    33428
dtype: int64

In [44]:
label_encoder = preprocessing.LabelEncoder() 
  
# Encode labels in column 'species'. 
enc_Src_Country= label_encoder.fit_transform(df[['country']]) 
New_cat_Src_Country = pd.DataFrame(enc_Src_Country)
New_cat_Src_Country= New_cat_Src_Country.rename(columns={0:'Src_country'})

In [45]:
New_cat_Src_Country.nunique()

Src_country    215
dtype: int64

### Ports

In [46]:
imp_ports = ['22', '443', '80', '53', '389', '25', '113', '123', '554', '520', '161', '995', '67', '993', '631', '110',
 '143', '0', '445', '137', '427', '138', '524', '514', '139', '1000', '784', '12', '465', '592', '587', '88', '2', '888', '21',
 '500', '544', '81', '418', '294', '34', '98', '68', '709', '23', '8', '625', '768', '579', '135', '104', '916', '877', '310',
 '490', '1', '82', '369', '1013', '83', '832', '843', '471', '118']

In [47]:
len(imp_ports)

64

In [48]:
df['ports']=np.where(df.Destination_Port.isin(imp_ports), df['Destination_Port'] , 'NaN')

In [49]:
df.ports.value_counts()

NaN     1085101
53       989469
80       250318
443       61556
123        2743
993        2316
110        1321
25         1214
995         953
113         378
22          331
21          321
161         258
143         113
587          71
427          60
465          44
81           34
631          27
843          24
389          16
524          15
139          14
500          11
888          10
23           10
82            7
118           4
1             3
1000          2
88            2
554           2
1013          1
8             1
592           1
544           1
916           1
Name: ports, dtype: int64

In [50]:
df = df[df.ports != 'NaN']

In [51]:
df

Duration Protocol       Source_IP Source_Port Direction  \
2           3.056586      tcp    147.32.86.89        4768  outgoing   
3           3.111769      tcp    147.32.86.89        4788  outgoing   
4           3.083411      tcp    147.32.86.89        4850  outgoing   
5           3.097288      tcp    147.32.86.89        4866  outgoing   
10        328.361664      tcp    147.32.84.59       49185  outgoing   
17          2.948305      tcp  41.188.145.202        2288  outgoing   
18          2.956415      tcp  41.188.145.202        2291  outgoing   
21          6.035653      tcp  41.188.145.202        2288  outgoing   
22          6.028175      tcp  41.188.145.202        2291  outgoing   
26          3.015055      tcp   69.231.198.54        4144  outgoing   
32          2.916540      tcp   69.231.198.54        4177  outgoing   
33          2.923881      tcp   147.32.86.141        4303  outgoing   
40       1713.705688      tcp    147.32.84.59       49157  outgoing   
41       1713.706299      tcp    147.32.84.59       49156  outgoing   
42       1713.303589      tcp    147.32.84.59       49159  outgoing   
43       1575.188477      tcp    147.32.84.59       49290  outgoing   
44       1575.188477      tcp    147.32.84.59       49291  outgoing   
45       1627.699951      tcp    147.32.84.59       49236  outgoing   
48       1644.362549      tcp    147.32.84.59       49255  outgoing   
51       1164.148438      tcp    147.32.84.59        1652  outgoing   
54          7.019550      tcp     87.96.93.53       50615  outgoing   
56        624.185059      tcp    147.32.84.94       38362  outgoing   
75          9.067471      tcp    178.23.90.30       57207  outgoing   
76          9.016273      tcp    178.23.90.30       57208  outgoing   
77        737.728821      tcp    147.32.84.59       52715  outgoing   
80        864.827332      tcp    147.32.84.59       49160  outgoing   
85          6.960091      tcp     87.96.93.53       51511  outgoing   
88          2.996404      tcp   85.235.176.38        4493  outgoing   
89          2.996822      tcp   85.235.176.38        4497  outgoing   
91          6.037753      tcp   85.235.176.38        4493  outgoing   
...              ...      ...             ...         ...       ...   
2824580     0.000340      udp   147.32.84.170       49913   two-way   
2824581     0.279334      udp   147.32.84.170       58369   two-way   
2824582     0.000402      udp   147.32.84.170       58721   two-way   
2824583     0.367496      tcp    147.32.85.76        1525  outgoing   
2824585     0.000518      udp    147.32.85.60       63858   two-way   
2824586     0.000569      udp    147.32.85.60       55839   two-way   
2824587     0.018506      tcp    147.32.85.60       52573  outgoing   
2824588     0.036315      tcp    147.32.85.60       52574  outgoing   
2824594     0.050707      tcp    147.32.85.60       52575  outgoing   
2824600     0.001184      udp  111.243.43.148       11994   two-way   
2824603     0.108535      tcp   66.249.72.141       62493  outgoing   
2824604     0.000290      udp    147.32.86.20       65160   two-way   
2824605     0.153318      tcp    147.32.86.20        4161  outgoing   
2824607     0.000373      udp    147.32.84.59       60428   two-way   
2824609     0.000333      udp    147.32.84.59       57875   two-way   
2824613     0.021536      tcp   66.249.66.241       40474  outgoing   
2824614     0.000233      udp   147.32.84.138       40049   two-way   
2824615     0.000256      udp   147.32.84.138       47007   two-way   
2824616     0.000299      udp   147.32.84.138       53567   two-way   
2824617     0.000303      udp   147.32.84.138       42937   two-way   
2824619     0.000318      udp   147.32.84.138       56632   two-way   
2824620     0.000138      udp   147.32.84.138       54724   two-way   
2824621     0.000209      udp   147.32.84.138       53255   two-way   
2824622     0.000253      udp   147.32.84.138       40427   two-way   
2824624     0.000144      udp   147.

# Merging everything

In [52]:
Interval_Variables = df[['Source_IP','Destination_IP','Duration','Packets','Total_Bytes','Source_Bytes','ports']]
Interval_Variables.head()

Source_IP Destination_IP    Duration  Packets  Total_Bytes  \
2   147.32.86.89    77.75.73.33    3.056586        3          182   
3   147.32.86.89    77.75.73.33    3.111769        3          182   
4   147.32.86.89    77.75.73.33    3.083411        3          182   
5   147.32.86.89    77.75.73.33    3.097288        3          182   
10  147.32.84.59    147.32.80.7  328.361664        7          760   

    Source_Bytes ports  
2            122    80  
3            122    80  
4            122    80  
5            122    80  
10           520    80

In [53]:
a = pd.merge(Interval_Variables, New_cat_flags, right_index=True,left_index=True)
b = pd.merge(a, New_cat_protocol, right_index=True,left_index=True)
c = pd.merge(b, New_cat_Src_City, right_index=True,left_index=True)
d = pd.merge(c, New_cat_dir, right_index=True,left_index=True)
final_X = pd.merge(d, New_cat_Src_Country, right_index=True,left_index=True)


In [54]:
final_X

Source_IP  Destination_IP     Duration  Packets  Total_Bytes  \
2          147.32.86.89     77.75.73.33     3.056586        3          182   
3          147.32.86.89     77.75.73.33     3.111769        3          182   
4          147.32.86.89     77.75.73.33     3.083411        3          182   
5          147.32.86.89     77.75.73.33     3.097288        3          182   
10         147.32.84.59     147.32.80.7   328.361664        7          760   
17       41.188.145.202   147.32.84.229     2.948305        3          184   
18       41.188.145.202   147.32.84.229     2.956415        3          184   
21       41.188.145.202   147.32.84.229     6.035653        3          184   
22       41.188.145.202   147.32.84.229     6.028175        3          184   
26        69.231.198.54   147.32.86.179     3.015055        3          184   
32        69.231.198.54   147.32.86.179     2.916540        3          184   
33        147.32.86.141    77.75.73.158     2.923881        3          186   
40         147.32.84.59     147.32.80.7  1713.705688       12         2636   
41         147.32.84.59     147.32.80.7  1713.706299       12         2636   
42         147.32.84.59  213.199.181.90  1713.303589       11         1080   
43         147.32.84.59  209.85.148.103  1575.188477        6          372   
44         147.32.84.59  209.85.148.103  1575.188477        6          372   
45         147.32.84.59  74.125.232.223  1627.699951       15         3250   
48         147.32.84.59  74.125.232.223  1644.362549       19         3938   
51         147.32.84.59    147.32.80.13  1164.148438        7          422   
54          87.96.93.53   147.32.84.229     7.019550       14          966   
56         147.32.84.94    147.32.80.13   624.185059       36        17444   
75         178.23.90.30   147.32.84.229     9.067471        7          430   
76         178.23.90.30   147.32.84.229     9.016273        7          430   
77         147.32.84.59  74.125.232.217   737.728821       21         4759   
80         147.32.84.59  213.199.181.90   864.827332       11         1080   
85          87.96.93.53   147.32.84.229     6.960091       14          966   
88        85.235.176.38   147.32.84.229     2.996404        3          184   
89        85.235.176.38   147.32.84.229     2.996822        3          184   
91        85.235.176.38   147.32.84.229     6.037753        3          184   
...                 ...             ...          ...      ...          ...   
2396698    147.32.84.59     77.78.99.23     0.182225       14         5183   
2396699   147.32.86.194  195.113.232.97   115.638954       12         3945   
2396700    147.32.84.59     147.32.80.9     0.000242        2          277   
2396701    147.32.84.59   209.85.148.96   122.630089       17         5264   
2396702    147.32.84.59   209.85.148.96    69.848991       17         4500   
2396705   147.32.86.194  173.230.139.22     0.296718       23        16052   
2396709    147.32.84.59  209.85.149.189    54.093918       19         4482   
2396711    147.32.84.59     147.32.80.9     0.118327        2          284   
2396712    147.32.84.59     147.32.80.9     0.000344        2          364   
2396714    147.32.84.59     147.32.80.9     0.000425        2          262   
2396715    147.32.84.59  74.125.232.222    69.716614       17         4341   
2396716    147.32.84.59  74.125.232.222   114.735611       17         4914   
2396723   66.249.66.242    147.32.87.38     2.460102       28        17978   
2396727    147.32.84.59   64.12.239.106     7.281869        9         1535   
2396735   147.32.86.192     77.75.72.46     0.084921       59        33080   
2396736    147.32.84.59     147.32.80.9     0.000290        2          204   
2396737    147.32.84.59     147.32.80.9     0.000390        2          230   
2396738    147.32.84.59     147.32.80.9     0.000328        2          292   
2396741   147.32.84.138     147.32.80.9     0.000164        2          214   
2396742   147.32.84.138     147.32.8

In [55]:
final_X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1114855 entries, 2 to 2396752
Data columns (total 12 columns):
Source_IP         1114855 non-null object
Destination_IP    1114855 non-null object
Duration          1114855 non-null float64
Packets           1114855 non-null int64
Total_Bytes       1114855 non-null int64
Source_Bytes      1114855 non-null int64
ports             1114855 non-null object
Flags             1114855 non-null int32
Protocol          1114855 non-null int32
Src_city          1114855 non-null int32
Direction         1114855 non-null int32
Src_country       1114855 non-null int32
dtypes: float64(1), int32(5), int64(3), object(3)
memory usage: 89.3+ MB


# Defining X and y variables
## Target

In [56]:
Target = df.iloc[:,-4:-3]
Target.rename(columns = {-4:'label'}, inplace = True) 
Target['Label'].value_counts()

Background    1253997
Normal          30144
Botnet          27511
Name: Label, dtype: int64

In [57]:
#Classifying labels into two categories
#botnet = Yes
#Background,Normal = No
a = ['Botnet']
Target['yes/no'] = np.where(Target['Label'].isin(a), 'yes', 'no')

In [58]:
Target['yes/no'].value_counts()

no     1284141
yes      27511
Name: yes/no, dtype: int64

## Downsampling

In [59]:
from sklearn.utils import resample

d = pd.merge(Target, final_X, right_index=True, left_index=True)
d.head()

Label yes/no     Source_IP Destination_IP    Duration  Packets  \
2   Background     no  147.32.86.89    77.75.73.33    3.056586        3   
3   Background     no  147.32.86.89    77.75.73.33    3.111769        3   
4   Background     no  147.32.86.89    77.75.73.33    3.083411        3   
5   Background     no  147.32.86.89    77.75.73.33    3.097288        3   
10  Background     no  147.32.84.59    147.32.80.7  328.361664        7   

    Total_Bytes  Source_Bytes ports  Flags  Protocol  Src_city  Direction  \
2           182           122    80    148         3     23410          1   
3           182           122    80    148         3     23410          1   
4           182           122    80    148         3     23410          1   
5           182           122    80    148         3     23410          1   
10          760           520    80    144         3     23410          1   

    Src_country  
2            50  
3            50  
4            50  
5            50  
10           50

In [60]:
d['yes/no'].value_counts()

no     1094340
yes      20515
Name: yes/no, dtype: int64

In [61]:
majo = d.loc[d['yes/no']=='no']
mino = d.loc[d['yes/no']=='yes']

In [62]:
# downsample majority class
maj_dsampled = resample(majo, 
                        replace=False,     # sample with replacement
                        n_samples=20515,   # to match minority class
                        random_state=1)    # reproducible results

In [63]:
dsampled = pd.concat([mino, maj_dsampled])
dsampled['yes/no'].value_counts()

yes    20515
no     20515
Name: yes/no, dtype: int64

## X and Y variables

In [64]:
y=dsampled[['yes/no']]
X=dsampled.drop(['yes/no','Label','Source_IP','Destination_IP'],axis=1)


X

Duration  Packets  Total_Bytes  Source_Bytes ports  Flags  \
675613     0.000278        2          203            64    53      4   
675948     0.020525        2          590            87    53      4   
675953     0.045125        7          882           629    80      4   
690001     0.336250        2          215            71    53      4   
691744     0.459301        2          212            77    53      4   
691823     0.216252       37        27990          1046    80      4   
691960     1.371909        3          295           156    53      4   
692271     0.313117        2          202            72    53      4   
692303    13.535808       70        41849          1962    80     45   
692584     0.547071        2          223            81    53      4   
692638     0.758234        9         2589           611   888      4   
692775     0.559377        2          218            77    53      4   
692825     7.599358      150       137436          3721    82     48   
693738     0.000355        2          223            81    53      4   
693739     0.980806       11         1320           833   888      4   
693891     0.131772        2          283            71    53     45   
693917     1.029718       11         1328           841   888      4   
693919   297.349731       11         1569           714    80    126   
694034     0.422243        2          182            71    53      4   
694064     2.807242       48        37425          1441    81      4   
694086     0.829001       19        10327           837    80      4   
694202     0.000417        2          202            72    53      4   
694203     7.151825       43        32509          1260    80      4   
694442     0.000345        2          223            81    53      4   
694443     0.990888       11         1324           837   888     45   
694508     0.251006       61        48359          1585    80      4   
694564     0.973053       11         1328           841   888      4   
694578     1.354025        6          366           186   443      4   
695014     9.560554        7          558           436   443     45   
695054     0.000299        2          182            71    53      4   
...             ...      ...          ...           ...   ...    ...   
2390125    0.266589       22         8589          1558    80      4   
1280857    0.273591       10         1821           863    80      4   
382308     0.000318        2          216            83    53      4   
2019823    0.000249        2          214            81    53      4   
838478     0.000332        2          314            85    53      4   
768337     0.000320        2          207            66    53      4   
1335448    0.000211        2          214            81    53    160   
1389584    0.000250        2          228            86    53      4   
923863     0.000336        2          214            81    53      4   
701923     0.000253        2          224            78    53      4   
152039     8.862285        6          390           132   443      4   
1011580    0.000168        2          214            81    53      4   
2223259    0.000577        2          214            81    53      4   
1886580    0.000167        2          214            82    53      4   
1655001    0.000287        2          214            81    53      4   
374626     0.000328        2          216            83    53     45   
477260     0.000449        2          312            77    53      4   
1240267    0.119287        2          242            81    53      4   
535449     0.001371        2          128            60   443      4   
1575991    0.000161        2          214            81    53      4   
336574     0.000240        2          214            81    53      4   
197592     0.000259        2          216            83    53      4   
597512     0.000290        2          214            81    53      4   
2000631    0.000342        2          214          

In [65]:
y

yes/no
675613     yes
675948     yes
675953     yes
690001     yes
691744     yes
691823     yes
691960     yes
692271     yes
692303     yes
692584     yes
692638     yes
692775     yes
692825     yes
693738     yes
693739     yes
693891     yes
693917     yes
693919     yes
694034     yes
694064     yes
694086     yes
694202     yes
694203     yes
694442     yes
694443     yes
694508     yes
694564     yes
694578     yes
695014     yes
695054     yes
...        ...
2390125     no
1280857     no
382308      no
2019823     no
838478      no
768337      no
1335448     no
1389584     no
923863      no
701923      no
152039      no
1011580     no
2223259     no
1886580     no
1655001     no
374626      no
477260      no
1240267     no
535449      no
1575991     no
336574      no
197592      no
597512      no
2000631     no
1347023     no
1700006     no
358949      no
904241      no
771349      no
715143      no

[41030 rows x 1 columns]

We have 41,030 records now.

# Modeling 

<font size=5><font color='blue'> The top 3 methods from the previous analysis are XGBoost, Decision Tree and Bagging Classifier.

## XGBoost

In [66]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.3,random_state=42)
scaler_minmax = MinMaxScaler()
scaler_minmax.fit(X_train)
scaler_minmax.fit(X_test)

X_train = scaler_minmax.transform(X_train)
X_test = scaler_minmax.transform(X_test)

In [67]:
#XGBoost
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score

# Fitting the model. 
model = XGBClassifier()
model.fit(X_train, y_train)

# Making predictions for the test data. 
y_pred = model.predict(X_test)
predictions = y_pred     # replaced [round(value) for value in y_pred] with y_pred

# Evaluating predictions for the test data. 
accuracy_score(y_test, predictions)

#The first time I ran, gradient boosting produced 92%, this is what is in the presentation

0.9558859371191811

## Decision Tree

In [68]:
#Decision Tree
from sklearn.tree import DecisionTreeClassifier
dtree = DecisionTreeClassifier(criterion='gini')
dtree.fit(X_train,y_train)
dtree.score(X_test, y_test)

0.9162401494841174

## Bagging Classifier

In [69]:
from sklearn.metrics import confusion_matrix as cm
from sklearn.ensemble import BaggingClassifier
clf = BaggingClassifier(n_estimators=10).fit(X_train, y_train)

clf.score(X_test, y_test)

0.9320009748964172

# Source city/country removed

In [71]:
X = dsampled.drop(['Label', 'yes/no', 'Source_IP', 'Destination_IP', 'Src_city', 'Src_country'], axis=1)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.3,random_state=42)
scaler_minmax = MinMaxScaler()
scaler_minmax.fit(X_train)
scaler_minmax.fit(X_test)

X_train = scaler_minmax.transform(X_train)
X_test = scaler_minmax.transform(X_test)

## XGBoost

In [72]:
#XGBoost
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score

# Fitting the model. 
model = XGBClassifier()
model.fit(X_train, y_train)

# Making predictions for the test data. 
y_pred = model.predict(X_test)
predictions = y_pred     # replaced [round(value) for value in y_pred] with y_pred

# Evaluating predictions for the test data. 
accuracy_score(y_test, predictions)


0.956129661223495

## Decision Tree

In [73]:
#Decision Tree
from sklearn.tree import DecisionTreeClassifier
dtree = DecisionTreeClassifier(criterion='gini')
dtree.fit(X_train,y_train)
dtree.score(X_test, y_test)

0.9246892517669998

## Bagging Classifier

In [75]:
from sklearn.metrics import confusion_matrix as cm
from sklearn.ensemble import BaggingClassifier
clf = BaggingClassifier(n_estimators=10).fit(X_train, y_train)

clf.score(X_test, y_test)

0.936550491510277